src/main/java/org/candlepin/subscriptions/subscription/SubscriptionSyncController.java

```diff
@@ -321,4 +337,51 @@ public void forceSyncSubscriptionsForOrg(String orgId) {
     var subscriptions = subscriptionService.getSubscriptionsByOrgId(orgId);
     subscriptions.forEach(this::syncSubscription);
   }
+
+  @Transactional
+  public List<org.candlepin.subscriptions.db.model.Subscription> findSubscriptionsAndSyncIfNeeded(
+      String accountNumber,
+      Optional<String> orgId,
+      Key usageKey,
+      OffsetDateTime rangeStart,
+      OffsetDateTime rangeEnd,
+      BillingProvider billingProvider) {
+    Assert.isTrue(Usage._ANY != usageKey.getUsage(), "Usage cannot be _ANY");
+    Assert.isTrue(ServiceLevel._ANY != usageKey.getSla(), "Service Level cannot be _ANY");
+
+    String productId = usageKey.getProductId();
+    Set<String> productNames = tagProfile.getOfferingProductNamesForTag(productId);
+    if (productNames.isEmpty()) {
+      log.warn("No product names configured for tag: {}", productId);
+      return Collections.emptyList();
+    }
+
+    List<org.candlepin.subscriptions.db.model.Subscription> result =
+        subscriptionRepository.findByAccountAndProductNameAndServiceLevel(
+            accountNumber, usageKey, productNames, rangeStart, rangeEnd, billingProvider);
+
+    if (result.isEmpty()) {
+      /* If we are missing the subscription, call out to the RhMarketplaceSubscriptionCollector
+      to fetch from Marketplace.  Sync all those subscriptions. Query again. */
+      if (orgId.isEmpty()) {
+        orgId = Optional.of(accountService.lookupOrgId(accountNumber));
+      }
+      log.info("Syncing subscriptions for account {} using orgId {}", accountNumber, orgId.get());
+      forceSyncSubscriptionsForOrg(orgId.get());
+      result =
+          subscriptionRepository.findByAccountAndProductNameAndServiceLevel(
+              accountNumber, usageKey, productNames, rangeStart, rangeEnd, billingProvider);
+    }
+
+    if (result.isEmpty()) {
+      log.error(
+          "No subscription found for account {} with key {} and product names {}",
+          accountNumber,
+          usageKey,
+          productNames);
+      return result;
```

Comment:
does this need to be here? won't result get returned anyway at line 385?

Reply:
Fixup incoming